go-controller/pkg/clustermanager/pod/ip_allocator.go

```diff
@@ -0,0 +1,265 @@
+package pod
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+
+	nettypes "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+
+	subnetipallocator "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/allocator/ip/subnet"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+// PodIPAllocator acts on pods events handed off by the cluster network
+// controller and allocates or releases IPs for them updating the pod annotation
+// as necessary with all the additional information derived from those IPs.
+type PodIPAllocator struct {
+	netInfo util.NetInfo
+
+	// allocator of IPs within subnets
+	allocator subnetipallocator.Interface
+
+	// track pods that have been released but not deleted yet so that we don't
+	// release more than once
+	releasedPods      map[string]sets.Set[string]
```

Comment:
not sure why this is needed. Cant you just search the informer cache for any other pod that is not completed that is not using the IP and then release it if not? (similar to what primary path does today)

I guess maybe the cost will be too high, cause the secondary IP is stored in the annotation and not in the pod status ips? So we will have to json unmarshal a bunch of times?